zookeeper-server/zookeeper-server-3.6.3/src/test/java/com/yahoo/vespa/zookeper/VespaZooKeeperTest.java

```diff
@@ -0,0 +1,256 @@
+                phaser.arriveAndAwaitAdvance();
+                while (config != null) {
+                    new ReconfigurableVespaZooKeeperServer(reconfigurer, config);
+                    phaser.arriveAndAwaitAdvance(); // server is now up, let test thread sync here
+                    phaser.arriveAndAwaitAdvance(); // wait before reconfig/teardown to let test thread do stuff
+                }
+                reconfigurer.deconstruct();
+            }));
+        }
+
+        void await() throws ExecutionException, InterruptedException, TimeoutException {
+            future.get().get(30, SECONDS);
+        }
+    }
+
+    static List<ZookeeperServerConfig> getConfigs(int removed, int retired, int active, int joining) {
+        return IntStream.rangeClosed(1, removed + retired + active)
+                        .mapToObj(id -> getConfig(removed, retired, active, joining, id))
+                        .collect(toList());
+    }
+
+    // Config for server #id among retired + active servers, of which the last may be joining, and with offset removed.
+    static ZookeeperServerConfig getConfig(int removed, int retired, int active, int joining, int id) {
+        if (id <= removed)
+            return null;
+
+        Path tempDir = tempDirRoot.resolve("zookeeper-" + id);
+        return new ZookeeperServerConfig.Builder()
+                .clientPort(getPorts(id).get(0))
+                .dataDir(tempDir.toString())
+                .zooKeeperConfigFile(tempDir.resolve("zookeeper.cfg").toString())
+                .myid(id)
+                .myidFile(tempDir.resolve("myid").toString())
+                .dynamicReconfiguration(true)
+                .server(IntStream.rangeClosed(removed + 1, removed + retired + active)
+                                 .mapToObj(i -> new ZookeeperServerConfig.Server.Builder()
+                                         .id(i)
+                                         .clientPort(getPorts(i).get(0))
+                                         .electionPort(getPorts(i).get(1))
+                                         .quorumPort(getPorts(i).get(2))
+                                         .hostname("localhost")
+                                         .joining(i - removed > retired + active - joining)
+                                         .retired(i - removed <= retired))
+                                 .collect(toList()))
+                .build();
+    }
+
+    static List<Integer> getPorts(int id) {
+        if (ports.size() < id * 3) {
+            int previousPort;
+            if (ports.isEmpty()) {
+                String[] version = System.getProperty("zk-version").split("\\.");
+                int versionPortOffset = 0;
+                for (String part : version)
+                    versionPortOffset = 32 * (versionPortOffset + Integer.parseInt(part));
+                previousPort = 20000 + versionPortOffset % 30000;
```

Comment:
This 20000 and 30000 has  something to do with the module port range Harald mentioned?  Extract to constants?

Reply:
Hmm, they're just picked from thin air, really. Well, those other modules prefer things below 20k, but staying below 50k is just by chance.